lib/discover_endpoint.js

```diff
@@ -0,0 +1,305 @@
+    endpointRequest.httpRequest.headers['x-amz-api-version'] = apiVersion;
+  }
+}
+
+/**
+ * If api call gets invalid endoint exception, SDK should attempt to remove the invalid
+ * endpoint from cache.
+ * @api private
+ */
+function invalidateCachedEndpoints(response) {
+  var error = response.error;
+  var httpResponse = response.httpResponse;
+  if (error &&
+    (error.code === 'InvalidEndpointException' || httpResponse.statusCode === 421)
+  ) {
+    var request = response.request;
+    var operations = request.service.api.operations || {};
+    var inputShape = operations[request.operation] ? operations[request.operation].input : undefined;
+    var identifiers = marshallCustomeIdentifiers(request, inputShape);
+    var cacheKey = getCacheKey(request);
+    cacheKey = util.update(cacheKey, identifiers);
+    AWS.endpointCache.remove(cacheKey);
+  }
+}
+
+/**
+ * If endpoint is explicitly configured, SDK should not do endpoint discovery in anytime.
+ * @param [object] clien Service client object.
+ * @api private
+ */
+function hasCustomEndpoint(client) {
+  //if set endpoint is set for specific client, enable endpoint discovery will raise an error.
+  if (client._originalConfig && client._originalConfig.endpoint && client._originalConfig.endpointDiscoveryEnabled === true) {
+    throw util.error(new Error(), {
+      code: 'ConfigurationException',
+      message: 'If custome endpoint is supplied, endpointDiscoverEnabled should be turned off.'
+    });
+  };
+  var svcConfig = AWS.config[client.serviceIdentifier] || {};
+  return AWS.config.endpoint || svcConfig.endpoint || (client._originalConfig && client._originalConfig.endpoint);
+}
+
+/**
+ * @api private
+ */
+function isFalsey(value) {
+  return ['false', '0'].indexOf(value) >= 0;
+}
+
+/**
+ * If endpoint discovery should perform for this request when endpoint discovery is optional.
+ * SDK performs config resolution in order like below:
+ * 1. If turned on client configuration(default to off) then turn on endpoint discovery.
+ * 2. If turned on in env AWS_ENABLE_ENDPOINT_DISCOVERY then turn on endpoint discovery.
+ * 3. If turned on in shared ini config file with key 'endpoint_discovery_enabled', then
+ *   turn on endpoint discovery.
+ * @param [object] request request object.
+ * @api private
+ */
+function isEndpointDiscoveryApplicable(request) {
+  var service = request.service || {};
+  if (service.config.endpointDiscoveryEnabled === true) return true;
+  if (Object.prototype.hasOwnProperty.call(process.env, endpointDiscoveryEnabledEnv)) {
+    if (process.env[endpointDiscoveryEnabledEnv] === undefined) {
+      throw util.error(new Error(), {
+        code: 'ConfigurationException',
+        message: 'environmental variable AWS_ENABLE_ENDPOINT_DISCOVERY cannot be set to nothing'
+      });
+    }
+    if (!isFalsey(process.env[endpointDiscoveryEnabledEnv])) return true;
+  }
+  //shared ini file is only available in Node
+  if (util.isBrowser()) return false;
```

Comment:
Sorry for the incoming tangent, but I have 2 thoughts here.
1. I think we should update `isBrowser` to no longer depend on `process` existing:
https://github.com/aws/aws-sdk-js/blob/29fc8d39b73a9c7c15b6bc8fb28cf5ef313691f4/lib/util.js#L39
We've heard reports of this breaking with Angular, and it doesn't really make sense that we're depending on a global that doesn't exist natively in browsers. Instead, I'd propose we update our (browser|react-native|node)_loader.js files to set the function to return true or false accordingly.

2. This is less important if we do the item above, but I wonder if it would make sense to separate this function. You could have 3 new functions, each in their own file that perform one of these checks. 
Then you could have a separate browser/node file that exports the same function. This function would import whatever checks they need. This has the benefit of your browser build not having to import the code used to check the shared Ini file, and you could then allow the shared ini check to import the shared-ini loader directly. However, it also means more source files and might be more difficult to track.